File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.java
import java.io.*;
import java.util.*;

class Main {
	static BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
	static BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
	static StringTokenizer st;
    public static void main(String[] args) throws Exception {
    	
    	st = new StringTokenizer(reader.readLine());
    	Node root = new Node(Integer.parseInt(st.nextToken()));
    	
    	while(true) {
    		String line;
            line = reader.readLine();
            if(line != null && !line.isEmpty() == true) {
            	int number = Integer.parseInt(line);
        		root.insert(number);
            } else {
            	break;
            }
    	}
    	
    	postorder(root);
    	writer.flush();
    	
    	
    }
    
    static void postorder(Node node) throws IOException {
    	if (node.leftChild != null) {
    		postorder(node.leftChild);
    	}
    	if (node.rightChild != null) {
    		postorder(node.rightChild);
    	}
    	writer.write(node.num + "\n");
    }
}

class Node {
	int num = -1;
	Node leftChild = null;
	Node rightChild = null;
	
	public Node(int n) {
		this.num = n;
	}
	
	public void insert(int n) {
		if (n < this.num) {
			if (this.leftChild == null) {
				this.leftChild = new Node(n);
			} else {
				this.leftChild.insert(n);
			}
		} else {
			if (this.rightChild == null) {
				this.rightChild = new Node(n);
			} else {
				this.rightChild.insert(n);
			}
		}
	}
}

File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.cc
#include <iostream>
#include <vector>

using namespace std;

class Node;
vector<Node> nodeList;

class Node {
  public:
    int num;
    int leftChild = -1;
    int rightChild = -1;
    Node(int n) : num(n) {}
};

void insert(int node, int num) {
  if (num > nodeList[node].num) {
    if (nodeList[node].rightChild == -1) {
      nodeList.emplace_back(Node(num));
      nodeList[node].rightChild = nodeList.size() - 1;
    } else {
      insert(nodeList[node].rightChild, num);
    }
  } else {
    if (nodeList[node].leftChild == -1) {
      nodeList.emplace_back(Node(num));
      nodeList[node].leftChild = nodeList.size() - 1;
    } else {
      insert(nodeList[node].leftChild, num);
    }
  }
}

void postorder(int n) {
  if (nodeList[n].leftChild != -1) postorder(nodeList[n].leftChild);
  if (nodeList[n].rightChild != -1) postorder(nodeList[n].rightChild);
  cout << nodeList[n].num << "\n";
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);

  int rootNum;
  cin >> rootNum;
  nodeList.emplace_back(Node(rootNum));

  int input;
  while (cin >> input) {
    insert(0, input);
  }
  
  postorder(0);
  return 0;
}